### PR TITLE
Response sorting

### LINKED ISSUE #1023: Response Requisition: Detail sort
Be nice to have sorting on the columns of the response req

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #1023 
Adds sorting using the same approach as requests

### CODE REVIEW

File: packages/requisitions/src/ResponseRequisition/DetailView/ContentArea.tsx [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
 import { useResponseLines, ResponseLineFragment } from '../api';
 
 interface ContentAreaProps {
   onRowClick: null | ((line: ResponseLineFragment) => void);
 }
 
 export const ContentArea = ({ onRowClick }: ContentAreaProps) => {
-  const { lines, onChangeSortBy, sortBy } = useResponseLines();
-  const columns = useResponseColumns({ sortBy, onChangeSortBy });
+  const { columns, lines } = useResponseLines();
```
Comment (mark-prins) [inline packages/requisitions/src/ResponseRequisition/DetailView/ContentArea.tsx:10]: not great - using columns from `useResponseColumns` won't work, after the column sort is changed the `useResponseLines` hook isn't rerendered and has the wrong sort value.
Just having another look at it..

Comment (mark-prins) [reply packages/requisitions/src/ResponseRequisition/DetailView/ContentArea.tsx:10]: I give up. fewer renders like this and it works, so I'll stick with it.

File: packages/requisitions/src/ResponseRequisition/DetailView/columns.ts [hunk 2]
```diff
@@ -1,45 +1,83 @@
...
+          rowData.linkedRequisitionLine?.itemStats?.availableStockOnHand,
+        getSortValue: rowData =>
+          rowData.linkedRequisitionLine?.itemStats?.availableStockOnHand ?? '',
+
+        label: 'label.customer-soh',
+        description: 'description.customer-soh',
+        width: 100,
+        align: ColumnAlign.Right,
+      },
       'requestedQuantity',
```
Comment (joshxg) [inline packages/requisitions/src/ResponseRequisition/DetailView/columns.ts:81]: this guy needs a `getSortValue` (I've looked into defaulting `getSortValue` to the `accessor`, but the return types are different)

File: packages/requisitions/src/ResponseRequisition/DetailView/ContentArea.tsx (hunk 1 above)
Comment (joshxg) [reply packages/requisitions/src/ResponseRequisition/DetailView/ContentArea.tsx:10]: Though I am kind of diggin the columns being all wrapped up nicely with the lines 🤔

File: packages/requisitions/src/ResponseRequisition/DetailView/columns.ts [hunk 3]
```diff
@@ -1,45 +1,83 @@
...
+const useSharedSortBy = () => {
+  const sharedSortBy = useStore();
+  const { sortBy, onChangeSortBy } = useSortBy<ResponseLineFragment>(
+    sharedSortBy.sortBy
+  );
+
+  useEffect(() => {
+    sharedSortBy.setSortBy(sortBy);
+  }, [sortBy]);
+  return { sortBy, onChangeSortBy };
```
Comment (joshxg) [inline packages/requisitions/src/ResponseRequisition/DetailView/columns.ts:32]: Tricky one!

1. Render data with columns [from content area hook]
1. Trigger sort by change [from lines hook]
2. Changes hook specific state [lines hook state]
3. Effect changes zustand state [global]
4. returns hook specific state [lines hook state]
5. Lines hook reruns as zustand state changed 
6. Lines hook see the line hook specific state which hasn't changed
7. data doesn't re-render as the columns and data haven't changed

Can return zustand state instead
`{ sortBy: sharedSortBy.sortBy }`

which seems to be working (but I like what you've done moving the hook, seems nice)

File: packages/requisitions/src/ResponseRequisition/api/hooks.ts [hunk 4]
```diff
@@ -105,22 +106,22 @@ export const useResponseFields = <
...
-    );
-  }, [sortBy, lines]);
-
-  return { lines: sorted, sortBy, onChangeSortBy };
+    const currentColumn = columns.find(({ key }) => key === sortBy.key);
+    const { getSortValue } = currentColumn ?? {};
+    return getSortValue
+      ? lines?.nodes.sort(getColumnSorter(getSortValue, !!sortBy.isDesc))
+      : lines?.nodes;
+  }, [sortBy.key, sortBy.isDesc, lines]);
```
Comment (joshxg) [inline packages/requisitions/src/ResponseRequisition/api/hooks.ts:122]: With this method of sorting, we re-order whenever the lines change, so when we open the modal and change a quantity, and we have sorted on quantity, the order changes. Is that good or bad?

**Review by joshxg — APPROVED** [review]
awesome, it's starting to look so real!

File: packages/requisitions/src/ResponseRequisition/api/hooks.ts (hunk 4 above)
Comment (mark-prins) [reply packages/requisitions/src/ResponseRequisition/api/hooks.ts:122]: In some ways good - you finish editing and the table is sorted according to the values shown.
In other ways bad - if you OK & Next then you'll get a fairly surprising next line and the nexting will stop before or after you expect it to.
Overall.. ok for now

File: packages/requisitions/src/ResponseRequisition/DetailView/columns.ts (hunk 3 above)
Comment (mark-prins) [reply packages/requisitions/src/ResponseRequisition/DetailView/columns.ts]: 🎉  thanks! I've shifted to use `sharedSortBy.sortBy` and that works nicely. now the `useColumns` hook can be used safely 😊

**Pull request merged**